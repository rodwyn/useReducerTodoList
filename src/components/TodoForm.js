import React from 'react';
import { useForm } from "../hooks/useForm";

export const TodoForm = ({handleAddTodo}) => {
  const [{description}, handleInputChange, reset] = useForm({
    description: ''
  });

  const handleSubmit = event => {
    event.preventDefault();
    
    if (description.trim().length < 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    };

    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Add ToDo</h4>
      <hr />
      <form onSubmit={ handleSubmit }>
        <input
          type="text"
          name="description"
          placeholder="Add Todo"
          autoComplete="off"
          className="form-control"
          onChange={handleInputChange}
          value={description}
            />
        <div className="d-grid gap-2 mt-2">
          <button className="btn btn-outline-primary" type="submit">Add</button>
        </div>
      </form>
    </>
  )
}
