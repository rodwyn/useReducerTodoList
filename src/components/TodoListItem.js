import React from 'react'

export const TodoListItem = ({todo, i, handleToggle, handleDelete}) => {
  return (
    <li
    className="list-group-item"
    key={todo.id}>
      <p
        className={`${todo.done && 'completed'}`}
        onClick={() => handleToggle(todo.id)}>{i + 1}.{todo.desc}</p>
      <button className="btn btn-danger" onClick={() => handleDelete(todo.id)}>Delete</button>
    </li> 
  )
}
