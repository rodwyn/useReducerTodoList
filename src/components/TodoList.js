import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleToggle, handleDelete}) => {
  return (
    <ul className="list-group list-group-flush">
      {
        todos.map((todo, i) => (
          <TodoListItem
            todo={todo}
            handleToggle={handleToggle}
            handleDelete={handleDelete} 
            i={i} 
            key={todo.id}
            />
        ))
      }
    </ul>
  )
}
