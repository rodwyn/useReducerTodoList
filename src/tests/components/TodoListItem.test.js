import React from 'react';
import {shallow} from 'enzyme';
import {TodoListItem} from '../../components/TodoListItem';
import {demosTodos} from '../fixtures/demosTodos';

describe('Testing <TodoListItem /> component.', () => {
  const handleToggle = jest.fn();
  const handleDelete = jest.fn();
  const todo = demosTodos[0];
  todo.done = true;
  const wrapper = shallow(
    <TodoListItem
      todo={demosTodos[0]}
      i={0}
      handleToggle={handleToggle}
      handleDelete={handleDelete}
      />
  );
  
  test('should render correctly.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call handleDelete function.', () => {
    wrapper.find('button').simulate('click');
    expect(handleDelete).toHaveBeenCalledWith(demosTodos[0].id);
    
  });

  test('should call handleToggle function.', () => {
    wrapper.find('p').simulate('click');
    expect(handleToggle).toHaveBeenCalledWith(demosTodos[0].id);
  })
  
  test('should render the text correctly.', () => {
    const paragraph = wrapper.find('p');
    
    expect(paragraph.text()).toBe(`1.${demosTodos[0].desc}`);
  })
  
  test('should have the complete className if ToDO.done is true.', () => {
    expect(wrapper.find('p').hasClass('completed')).toBe(true);
  });
});
