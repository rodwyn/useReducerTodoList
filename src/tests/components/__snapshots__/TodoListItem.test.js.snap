// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <TodoListItem /> component. should render correctly. 1`] = `
<li
  className="list-group-item"
  key="123"
>
  <p
    className="completed"
    onClick={[Function]}
  >
    1
    .
    Leran to walk.
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
